text/0000-type-changing-struct-update-syntax.md

```diff
@@ -0,0 +1,651 @@
+    state: S,
+    common_field1: &'static str,
+    common_field2: i32,
+}
+
+struct State1;
+struct State2;
+
+impl Machine<State1> {
+    fn into_state2(self) -> Machine<State2> {
+        // do stuff
+        Machine {
+            state: State2,
+            common_field1: self.common_field1,
+            common_field2: self.common_field2,
+        }
+    }
+}
+```
+
+It would be much more convenient to be able to write
+
+```rust
+Machine {
+    state: State2,
+    ..self
+}
+```
+
+instead of
+
+```rust
+Machine {
+    state: State2,
+    common_field1: self.common_field1,
+    common_field2: self.common_field2,
+}
+```
+
+but this is not possible in current Rust because `Machine<State1>` and
+`Machine<State2>` are different types even though they are both the `Machine`
+struct.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+It's often useful to create a new instance of a struct that uses most of an old
+instance's values but changes some. You can do this using struct update syntax.
+
+Consider a `User` type that can be in either the `LoggedIn` state or the
+`LoggedOut` state and has a few additional fields describing the properties of
+the user.
+
+```rust
+struct User<S> {
+    state: S,
+    email: String,
+    username: String,
+}
+
+struct LoggedIn;
+struct LoggedOut;
+```
+
+Let's say we have a logged-out user:
+
+```rust
+let logged_out = User {
+    state: LoggedOut,
+    email: String::from("ferris@example.com"),
+    username: String::from("ferris"),
+};
+```
+
+This example shows how we create a new `User` instance named `logged_in`
+without the update syntax. We set a new value for `state` but move the values
+of the other fields from `logged_out`.
+
+```rust
+let logged_in = User {
+    state: LoggedIn,
+    email: logged_out.email,
+    username: logged_out.username,
+};
+```
+
+Using struct update syntax, we can achieve the same effect more concisely, as
+shown below. The syntax `..` specifies that the remaining fields not explicitly
+set should be moved from the fields of the base instance.
+
+```rust
+let logged_in = User {
+    state: LoggedIn,
+    ..logged_out
+};
+```
+
+Note that the expression following the `..` is an *expression*; it doesn't have
+to be just an identifier of an existing instance. For example, it's often
+useful to use struct update syntax with `..Default::default()` to override a
+few field values from their default.
+
+Struct update syntax is permitted for instances of the *same* struct (`User` in
+the examples), even if they have different types (`User<LoggedOut>` and
+`User<LoggedIn>` in the examples) due to generic type or lifetime parameters.
+However, the types of the fields in the updated instance that are not
+explicitly listed (i.e. those that are moved with the `..` syntax) must be
+subtypes of the corresponding fields in the base instance, and all of the
+fields must be visible ([RFC 736]). In other words, the types of fields that
```

Comment:
Yeah so I guess per @scottmcm's point, maybe we should roll back this rule. I guess it's worth deciding that separately.